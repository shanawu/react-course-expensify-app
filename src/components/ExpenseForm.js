import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";

const now = moment();

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      // if there is an existing value ? yes - take the existing value no - default value
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : now,
      calendarFocused: false,
      errorMessage: "",
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    // !amount > when the amount is empty OR when the amount matches /^\d{1,}(\.\d{0,2})?$/
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    } // to prevent the date from being completely cleared
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault(); // to prevent the page from refreshing

    // if there is no value for description OR amount
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        errorMessage: "Please provide description and/or amount.",
      }));
    } else {
      this.setState(() => ({ errorMessage: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };
  render() {
    // {this.state.errorMessage && <p>{this.state.errorMessage}</p>} the errorMessage will only show when this.state.errorMessage exists
    return (
      <div>
        {this.state.errorMessage && (
          <p className="error-Message">{this.state.errorMessage}</p>
        )}
        <form onSubmit={this.onSubmit} data-testid="expense-form">
          <input
            type="text"
            placeholder="Description"
            autoFocus={true}
            value={this.state.description}
            onChange={this.onDescriptionChange}
            data-testid="description-input"
          ></input>
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
            data-testid="amount-input"
          ></input>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            id="date"
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
            data-testid="note-textarea"
          ></textarea>
          <button data-testid="expense-form-submit">Add a expense</button>
        </form>
      </div>
    );
  }
}
