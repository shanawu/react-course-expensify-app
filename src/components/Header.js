import React from "react";
import { NavLink } from "react-router-dom";

const Header = () => {
  return (
    <header>
      <h1>Expensify</h1>
      <NavLink
        to="/"
        className={({ isActive }) => (isActive ? "active-link" : "")}
      >
        Home
      </NavLink>
      <NavLink
        to="/create"
        className={({ isActive }) => (isActive ? "active-link" : "")}
      >
        Add Expense
      </NavLink>
      <NavLink
        to="/edit"
        className={({ isActive }) => (isActive ? "active-link" : "")}
      >
        Edit Expense
      </NavLink>
      <NavLink
        to="/help"
        className={({ isActive }) => (isActive ? "active-link" : "")}
      >
        Help
      </NavLink>
    </header>
  );
};

export default Header;
