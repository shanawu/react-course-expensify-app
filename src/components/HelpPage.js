import React from "react";

const HelpPage = () => {
  return <div>This is from the help component</div>;
};

export default HelpPage;
