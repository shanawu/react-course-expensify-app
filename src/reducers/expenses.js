// Expense Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    // add expense
    case "ADD_EXPENSE":
      return [...state, action.expense]; // spread operator to add the newly added expense without modifying the original value
    // remove expense
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id); // state.filter((expense) => expense.id !== action.expense.id)
    // edit expense
    case "EDIT_EXPENSE":
      return state.map((expense) => {
        if (expense.id === action.id) {
          // finding if there is a match in the expenses
          return {
            ...expense, // spread the expense
            ...action.updates, // spread the updates
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};
