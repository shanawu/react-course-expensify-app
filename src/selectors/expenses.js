import moment from "moment";
// Get Filtered Expenses

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;
      const textMatch =
        typeof text === undefined ||
        expense.description.toLowerCase().indexOf(text.toLowerCase()) >= 0;
      // typeof text === undefined || expense.description.toLowerCase().includes(text.toLowerCase())

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1; // 1: b come first, -1: a come first
      } else if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};
