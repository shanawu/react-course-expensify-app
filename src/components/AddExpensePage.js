import React from "react";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";

const AddExpensePage = (props) => {
  const navigate = useNavigate();
  const handleExpenseSubmit = (expense) => {
    console.log("Submitting expense:", expense); // Add this line

    props.dispatch(addExpense(expense));
    navigate("/");
  };

  return (
    <div>
      <h1>Add Expense</h1>
      <ExpenseForm onSubmit={handleExpenseSubmit} />
    </div>
  );
};

export default connect()(AddExpensePage);
