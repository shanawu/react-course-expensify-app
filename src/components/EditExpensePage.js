import React from "react";
import { useNavigate } from "react-router-dom";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

const EditExpensePage = (props) => {
  const navigate = useNavigate();
  const handleExpenseEdit = (expense) => {
    props.dispatch(editExpense(props.expense.id, expense));
    navigate("/");
  };
  const handleRemoveExpense = (expense) => {
    props.dispatch(removeExpense({ id: props.expense.id }));
    navigate("/");
  };

  return (
    <div>
      <h1>Edit Expense</h1>
      <ExpenseForm expense={props.expense} onSubmit={handleExpenseEdit} />
      <button data-testid="remove-button" onClick={handleRemoveExpense}>
        Remove
      </button>
    </div>
  );
};

// finding the matched state and pass it to props
const mapStateToProps = (state, props) => {
  const params = { id: window.location.pathname.split("/")[2] };
  return {
    expense: state.expenses.find((expense) => expense.id === params.id),
  };
};

export default connect(mapStateToProps)(EditExpensePage);
