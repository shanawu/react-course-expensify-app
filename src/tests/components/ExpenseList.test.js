import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("should render ExpenseList with expenses", () => {
  const { container } = render(
    <BrowserRouter basename="/">
      <ExpenseList expenses={expenses} />
    </BrowserRouter>
  );

  expect(container).toMatchSnapshot();
});

test("should render ExpenseList component with no expenses", () => {
  const { container } = render(
    <BrowserRouter basename="/">
      <ExpenseList expenses={[]} />
    </BrowserRouter>
  );
  expect(container).toMatchSnapshot();
});
