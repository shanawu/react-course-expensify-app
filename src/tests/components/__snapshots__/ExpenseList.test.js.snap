// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList component with no expenses 1`] = `
<div>
  <div>
    <h1>
      Expense List
    </h1>
    <p>
      No expenses
    </p>
  </div>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <div>
    <h1>
      Expense List
    </h1>
    <div>
      <a
        href="/edit/1"
      >
        Gas
      </a>
      <p>
        156
         - 
        0
      </p>
    </div>
    <div>
      <a
        href="/edit/2"
      >
        Rent
      </a>
      <p>
        109500
         - 
        -345600000
      </p>
    </div>
    <div>
      <a
        href="/edit/3"
      >
        Groceries
      </a>
      <p>
        569
         - 
        345600000
      </p>
    </div>
  </div>
</div>
`;
